### DISCUSSION TITLE
Is it possible override FileNavigatorTree in Explorer

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-theia/theia#14023] [General] [answered]
Hi,
    As we know, the explorer can display a directory tree, and the file is the latest level and can not be expanded.
I need to display the function node of a C++ file, so I developed an extension to display and create my custom CFileNode, as you can in follow picture on the right panel.
![image](https://github.com/user-attachments/assets/c914da0c-356a-4e04-9d43-67a27d90694a)
```
export function createCodeExplainContainer(parent: interfaces.Container): Container {
    const child = createFileTreeContainer(parent);

    child.unbind(FileTree);

    child.bind(CFileTree).toSelf();
    child.rebind(Tree).toService(CFileTree);

    child.unbind(FileTreeWidget);
    child.bind(CodeExplainWidget).toSelf();

    return child;
}
```
but I find it is too tedious to have two extensions that have the same tree. and if I can directly override FileNavigatorTree and use my custom CFileTree, the C++ file can expand child node in explorer, but I have not found the trick on how to do it.
is it possible  override FileNavigatorTree in Explorer?

**Comment by msujew** [comment, marked as answer]
Hey @wss29,

you need to override the whole child container and rebind the binding to `FileNavigatorWidget`. See:

https://github.com/eclipse-theia/theia/blob/f7e2ef91b0d1436f313b385cc3a961eae1dcddc4/packages/navigator/src/browser/navigator-container.ts#L34-L48

**Reply by wss29** [reply]
Hi @msujew
 Thanks for you replay. it work now
```
export function createCFileNavigatorContainer(parent: interfaces.Container): Container {
    const child = createFileTreeContainer(parent, {
        tree: CFileTree,
        model: FileNavigatorModel,
        widget: CodeExplainWidget,
        decoratorService: NavigatorDecoratorService,
        props: FILE_NAVIGATOR_PROPS,
    });

    return child;
}

export function reCreateFileNavigatorWidget(parent: interfaces.Container): FileNavigatorWidget {
    const child = createCFileNavigatorContainer(parent);
    child.bind(FileNavigatorTree).toService(CFileTree);
    child.bind(FileNavigatorWidget).toService(CodeExplainWidget);
    return child.get(FileNavigatorWidget);
}

    rebind(FileNavigatorWidget).toDynamicValue(ctx => reCreateFileNavigatorWidget(ctx.container));
```
